src/backend/vault-hsm/dh-curve25519.c

```diff
@@ -0,0 +1,131 @@
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "internal.h"
+#include "crypto/ed25519/ed25519.h"
+#include <string.h>
+
+/* We use ed25519's faster curved25519_scalarmult_basepoint() function
+   when deriving a public key from a private key.  Unfortunately ed25519
+   doesn't have an equivalent function for general curve25519 calculations
+   so we fall back to the curve25519-donna implementation for that. */
+
+int curve25519_donna(uint8_t *mypublic, const uint8_t *secret, const uint8_t *basepoint);
+
+typedef struct
+{
+    struct NoiseDHState_s parent;
+    uint8_t private_key[32];
+    uint8_t public_key[32];
+
+} NoiseCurve25519State;
+
+static int noise_curve25519_generate_keypair
+    (NoiseDHState *state, const NoiseDHState *other)
+{
+    NoiseCurve25519State *st = (NoiseCurve25519State *)state;
+    noise_rand_bytes(st->private_key, 32);
+    st->private_key[0] &= 0xF8;
+    st->private_key[31] = (st->private_key[31] & 0x7F) | 0x40;
+    curved25519_scalarmult_basepoint(st->public_key, st->private_key);
+    return NOISE_ERROR_NONE;
+}
+
+static int noise_curve25519_set_keypair
+        (NoiseDHState *state, const uint8_t *private_key,
+         const uint8_t *public_key)
+{
+    /* Check that the public key actually corresponds to the private key */
+    NoiseCurve25519State *st = (NoiseCurve25519State *)state;
+    uint8_t temp[32];
+    int equal;
+    curved25519_scalarmult_basepoint(temp, private_key);
+    equal = noise_is_equal(temp, public_key, 32);
+    memcpy(st->private_key, private_key, 32);
+    memcpy(st->public_key, public_key, 32);
+    return NOISE_ERROR_INVALID_PUBLIC_KEY & (equal - 1);
+}
+
+static int noise_curve25519_set_keypair_private
+        (NoiseDHState *state, const uint8_t *private_key)
+{
+    NoiseCurve25519State *st = (NoiseCurve25519State *)state;
+    memcpy(st->private_key, private_key, 32);
+    curved25519_scalarmult_basepoint(st->public_key, st->private_key);
+    return NOISE_ERROR_NONE;
+}
+
+static int noise_curve25519_validate_public_key
+        (const NoiseDHState *state, const uint8_t *public_key)
+{
+    /* Nothing to do here yet */
+    return NOISE_ERROR_NONE;
+}
+
+static int noise_curve25519_copy
+    (NoiseDHState *state, const NoiseDHState *from, const NoiseDHState *other)
```

Comment:
Since `other` is not used, shan't we check that it is set to NULL and return an error if it's not?

Reply:
Copy pasted noise-c code